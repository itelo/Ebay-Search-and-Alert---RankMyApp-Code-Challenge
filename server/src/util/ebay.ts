import fetch from "node-fetch";

function findItemsByKeywords(keyword: string) {
  return fetch(
    `https://svcs.sandbox.ebay.com/services/search/FindingService/v1?SERVICE-NAME=FindingService&OPERATION-NAME=findItemsByKeywords&SERVICE-VERSION=1.12.0&SECURITY-APPNAME=${
      process.env.EBAY_APP_ID
    }&RESPONSE-DATA-FORMAT=JSON&REST-PAYLOAD&GLOBAL-ID=EBAY-US&keywords=${keyword}&paginationInput.entriesPerPage=3&paginationInput.pageNumber=1&SortOrderType=PricePlusShippingLowest`
  ).then(r => r.json());
}

type Item = {
  _id: string;
  price: string;
  itemURL: string;
  title: string;
  imageURL: string;
};

export const findItemsBySearchPhrase = (searchPhrase: string) => {
  return new Promise<Item[]>((resolve, reject) => {
    findItemsByKeywords(searchPhrase).then(
      (data: any) => {
        // console.log(data.findItemsByKeywordsResponse[0].searchResult[0]);
        let result = [] as Item[];
        if (
          data &&
          data.findItemsByKeywordsResponse &&
          data.findItemsByKeywordsResponse[0] &&
          data.findItemsByKeywordsResponse[0].searchResult &&
          data.findItemsByKeywordsResponse[0].searchResult[0] &&
          data.findItemsByKeywordsResponse[0].searchResult[0].item
        ) {
          result = data.findItemsByKeywordsResponse[0].searchResult[0].item.map(
            (item: any) => {
              return {
                price: item.sellingStatus[0].currentPrice[0].__value__,
                itemURL: item.viewItemURL[0],
                _id: item.itemId[0],
                title: item.title[0],
                imageURL:
                  item.galleryURL && item.galleryURL[0]
                    ? item.galleryURL && item.galleryURL[0]
                    : undefined
              };
            }
          ) as Item[];
        }

        resolve(result);
      },
      (error: any) => {
        reject(error);
      }
    );
  });
};
