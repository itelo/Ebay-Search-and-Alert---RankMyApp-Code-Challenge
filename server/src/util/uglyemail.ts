type Item = {
  price: string;
  itemURL: string;
  title: string;
  imageURL: string;
};

export const uglyEbayItem = (item: Item) => {
  const { price, itemURL, title, imageURL } = item;

  const image = imageURL
    ? `<td><img src="${imageURL}" alt="" width="96" height="96" style="display: block;" /></td>`
    : "";
  return `
    <tr>
      <td width="260" valign="top">
        <table border="0" cellpadding="0" cellspacing="0" width="100%">
          <tr style="cursor:pointer"
            onclick="window.open('${itemURL}')">
            ${image}
            <td style="padding: 0 0 0 0;">
              <table border="0" cellpadding="0" cellspacing="0">
                <tr>
                  <td style="padding: 0 0 0 0;">
                    ${title}
                  </td>
                <tr>
                  <td style="padding: 0 0 0 0;">
                    US $${price}
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>`;
};
