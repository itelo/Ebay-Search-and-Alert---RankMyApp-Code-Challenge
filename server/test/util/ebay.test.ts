import { findItemsBySearchPhrase } from "../../src/util/ebay";
import dotenv from "dotenv";
dotenv.config({ path: ".env" });

describe("findItemsBySearchPhrase", () => {
  it("should return result for 'iphone'", async () => {
    try {
      const items = await findItemsBySearchPhrase("iphone");
      expect(items).toMatchSnapshot();
    } catch (err) {
      throw err;
    }
  });

  it("should return result for empty string", async () => {
    try {
      const items = await findItemsBySearchPhrase("");
      expect(items).toMatchSnapshot();
    } catch (err) {
      throw err;
    }
  });

  it("should return result for 'pirulito'", async () => {
    try {
      const items = await findItemsBySearchPhrase("pirulito");
      expect(items).toMatchSnapshot();
    } catch (err) {
      throw err;
    }
  });
});
