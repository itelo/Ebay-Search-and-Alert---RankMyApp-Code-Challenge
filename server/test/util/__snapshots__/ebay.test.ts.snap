// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findItemsBySearchPhrase should return result for 'iphone' 1`] = `
Array [
  Object {
    "_id": "110421025574",
    "imageURL": undefined,
    "itemURL": "http://cgi.sandbox.ebay.com/iPhone-Case-/110421025574",
    "price": "1.0",
    "title": "iPhone Case",
  },
  Object {
    "_id": "110415569962",
    "imageURL": "http://thumbs3.sandbox.ebaystatic.com/m/mu9EzEB3ehEMpsoBeU-HVIA/140.jpg",
    "itemURL": "http://cgi.sandbox.ebay.com/Apple-IPhone-X-64GB-Factory-Unlocked-4G-LTE-IOS-WiFi-12MP-Camera-Smartphone-/110415569962",
    "price": "593.95",
    "title": "Apple IPhone X 64GB \\"Factory Unlocked\\" 4G LTE IOS WiFi 12MP Camera Smartphone",
  },
  Object {
    "_id": "110418362389",
    "imageURL": "http://thumbs2.sandbox.ebaystatic.com/m/mHBltWTgpUAzBRdOnWiZzfg/140.jpg",
    "itemURL": "http://cgi.sandbox.ebay.com/Apple-iPhone-SE-16GB-64GB-Unlocked-Various-Colours-Model-A1723-/110418362389",
    "price": "74.99",
    "title": "Apple iPhone SE 16GB 64GB Unlocked, Various Colours Model A1723",
  },
]
`;

exports[`findItemsBySearchPhrase should return result for 'pirulito' 1`] = `Array []`;

exports[`findItemsBySearchPhrase should return result for empty string 1`] = `Array []`;
